pkg/liveshare/session.go

```diff
@@ -46,17 +52,41 @@ func (s *Session) registerRequestHandler(requestType string, h handler) func() {
 	return s.rpc.register(requestType, h)
 }
 
-// StartsSSHServer starts an SSH server in the container, installing sshd if necessary,
-// and returns the port on which it listens and the user name clients should provide.
-func (s *Session) StartSSHServer(ctx context.Context) (int, string, error) {
+func WithUserPublicKeyFile(path string) func(opts *StartSSHServerOptions) {
+	return func(opts *StartSSHServerOptions) {
+		opts.userPublicKeyFile = path
+	}
+}
+
+// StartSSHServer starts an SSH server in the container, installing sshd if necessary, applies specified
+// options, and returns the port on which it listens and the user name clients should provide.
+func (s *Session) StartSSHServer(ctx context.Context, configureOptions ...func(*StartSSHServerOptions)) (int, string, error) {
+	var params struct {
+		UserPublicKey string `json:"userPublicKey"`
+	}
+
 	var response struct {
 		Result     bool   `json:"result"`
 		ServerPort string `json:"serverPort"`
 		User       string `json:"user"`
 		Message    string `json:"message"`
 	}
 
-	if err := s.rpc.do(ctx, "ISshServerHostService.startRemoteServer", []string{}, &response); err != nil {
+	opts := StartSSHServerOptions{}
+	for _, configure := range configureOptions {
+		configure(&opts)
```

Comment:
What's the advantage of taking functions as input to `StartSSHServer`?

No matter how we slice it, we need to flatten a range of options. We could replace `WithUserPublicKeyFile` with a private helper to flatten a range of `StartSSHServerOptions` and take `sshOptions ...*StartSSHServerOptions` instead of `configureOptions ...func(*StartSSHServerOptions))`. This would make for simpler types in the other files.

Reply:
This is a great question, and there are many ways to make this same thing work as you've pointed out, so let me explain why I chose this option *buh dum tss*. 

- Option 1: pass in a single StartSshServerOptions object directly
  - `StartSshServer(options StartSshServerOptions)`
  - Pros: 
     - when you need to set options this is clean and straightforward on the caller side
  - Cons:
     - if you don't need to set options, you force the caller to pass in an empty options or `nil`
     - you also need to make all the properties of the options public, which makes it harder to change the interface later as callers will rely on direct access
- Option 2: helper methods which create partial options and squash them
   - `StartSshServer(options ...StartSshServerOptions)`
   - `WithPublicKeyPath(path string) StartSshServerOptions`
   - Pros: 
     - if a caller doesn't care about options they don't need to pass anything
     - you can keep all the properties private
   - Cons:
     - the squashing logic gets tedious as properties are added as you have to individually squash each one
- Option 3: configuration functions (current approach)
   - `StartSshServer(configureOptions ...func(*StartSshServerOptions))`
   - Pros:
     - same as Option 2
     - additionally you don't need squashing logic as you just keep updating a single options value
   - Cons:
     - the method signature is a bit unwieldy (although that can be improved with custom types)
- Option 4: configuration methods on the options itself
   - `StartSshServer(options StartSshServerOptions)`
   - `func (o *StartSshServerOptions) WithPublicKeyFile(path string)`
   - Pros:
      - properties can be private again
      - the call site is clean if you want to pass options (same as Option 1)
   - Cons:
      - also same as Option 1, if you don't want to pass options you still have to send a default or `nil`

There are even more options than these even. For example, `StartSshServer` and `StartSshServerWithOptions` can hide the `nil` options if callers don't want it. 

Truthfully, for this exact case it's small and specific enough that all options are truthfully fine and workable. I chose Option 3 because it had nice pros with few cons as well as sharing a pattern with the ssh keygen helper.